### ISSUE TITLE
Protein fitter doesn't run in MPI 

### ISSUE THREAD

**Issue by sjspielman** [veg/hyphy#537]
The new line of code introduced in [this recent commit](https://github.com/veg/hyphy/commit/53945ee02be8b3276c0c6d6aaa3d6e38f460b0eb) ("Fixing the issue that would prevent model fitting steps from reusing branch lengths") appears to have introduced a bug that prevents ProteinGTRFitter from running w/ HYPHYMPI (no bug with HYPHYMP).

Here is the call and the output w/ error messages. The first line that trips the error is the new line added with aforementioned commit.
```
[sjspielman@silverback validate]$ (echo /home/sjspielman/protfit/validate/simulations_nohet.list.1-0.1-1-1; echo 1; echo 1; echo 1;) | /opt/scyld/openmpi/1.6.3/gnu/bin/mpirun -hostfile ./hostfile -np 10 /home/sjspielman/repos-for-hyphy/hyphy/HYPHYMPI LIBPATH=/home/sjspielman/repos-for-hyphy/hyphy/res /home/sjspielman/repos-for-hyphy/hyphy/res/TemplateBatchFiles/SelectionAnalyses/ProtGTR_Fitter/ProteinGTRFit.bf

Analysis Description
--------------------
Fit a general time reversible model to a collection of training protein
sequence alignments. Generate substitution and scoring matrices
following the procedures described in Nickle et al 2007

- __Citation__: Nickle DC, Heath L, Jensen MA, Gilbert PB, Mullins JI, Kosakovsky Pond
SL (2007) HIV-Specific Probabilistic Models of Protein Evolution. PLoS
ONE 2(6): e503. doi:10.1371/journal.pone.0000503

- __Written by__: Sergei L Kosakovsky Pond and Stephanie J Spielman

- __Contact Information__: {spond,stephanie.spielman}@temple.edu

- __Analysis Version__: 0.01


Supply a list of files to include in the analysis (one per line) (`/home/sjspielman/repos-for-hyphy/hyphy/res/TemplateBatchFiles/SelectionAnalyses/ProtGTR_Fitter/`) 

			+----------------------------------------------------------------------------------------------+
			|A cache file of a prior analysis on this list of files was detected. Would you like to use it?|
			+----------------------------------------------------------------------------------------------+


	(1):[YES] Resume analysis using the detected cache file.
	(2):[NO] Launch a new analysis and *overwrite* the detected cache file.

 Please choose an option (or press q to cancel selection):

### Initial branch length fit


[PHASE 1] Performing initial branch length optimization using 0

### Loaded cached results for '/home/sjspielman/protfit/simulate/hyphy_input/simulated_lg100.dat' 1/3. Log(L) = 0
Loaded cached results for '/home/sjspielman/protfit/simulate/hyphy_input/simulated_lg10.dat' 2/3. Log(L) = 0
Loaded cached results for '/home/sjspielman/protfit/simulate/hyphy_input/simulated_lg1.dat' 3/3. Log(L) = 0
Overall Log(L) = 0


[PHASE 2] Performing initial REV fit to the data

### Fitting the REV model using constrained branch lengths proportions (REV-Phase-0)
Error:

Master node received an error:Illegal right hand side in call to Tree id = ...; it must be a string, a Newick tree spec or a topology

Function call stack
1 : Build Tree vZDdozcH.tree_0 from string null
-------
2 : ExecuteCommands in string "UseModel ("+model.ApplyModelToTree.modelID["id"]+");
						  Tree `id` = "+tree["string"]+";
						  " using basepath /home/sjspielman/repos-for-hyphy/hyphy/res/TemplateBatchFiles/libv3/models/.
-------
3 : model.ApplyModelToTree(Eval("&`lf_components[2*i + 1]`"),tree[i],{"default":user_model},None)
-------
4 : protein_gtr.rev.mle=estimators.FitSingleModel_Ext(utility.Map(filter_info,"_value_","_value_['name']"),trees,protein_gtr.rev_model_branch_lengths,initial_values,{"proportional-branch-length-scaler":proportional_scalers})
-------
5 : protein_gtr.current_gtr_fit=protein_gtr.fitGTRtoFileList(utility.Map(utility.Filter(protein_gtr.analysis_results,"_value_","_value_/'"+protein_gtr.phase_key+"'"),"_value_","_value_['"+protein_gtr.phase_key+"']"),protein_gtr.current_gtr_fit,result_key,FALSE)
-------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD 
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
```

**Comment by sjspielman** [comment]
Spoke too soon, a different bug keeps cropping up (stochastically, with a large number of data files. have 3 data files for example does not trigger bug that I've seen) in HYPHYMP as well.  I'm attaching here the stdout and stderr from the silverback run (o.validate_mp and e.validate_mp), as well as the relevant messages.log.
[e.validate_mp.txt](https://github.com/veg/hyphy/files/876183/e.validate_mp.txt)
[o.validate_mp.txt](https://github.com/veg/hyphy/files/876184/o.validate_mp.txt)
[messages.log.txt](https://github.com/veg/hyphy/files/876189/messages.log.txt)

**Comment by spond** [comment]
My least favorite kind of bug -- stochastic, long time to recreate, and memory related :(

**Comment by sjspielman** [comment]
Happy to brighten your day! ;)

**Comment by spond** [comment]
Dear @sjspielman,

Please check to see if f53f9daa32c9c7ab84d9437d6d54b8a6cb6859ef fixes this issue. It may also fix the remaining MP issue. The stochastic nature of this error makes sense because it was likely caused by incorrect memory management. 

Best,
Sergei

**Comment by sjspielman** [comment]
Still waiting on MP to see what happens, but some kind of bug indeed remains in MPI. Attaching here relevant errors and messages files. The error appears to have occurred *while* fitting the initial REV model to data ("Phase 2").
```
Master node received an error:Optimization routine error

Function call stack
1 : Optimize storing into, mles, the following likelihood function:likelihoodFunction ; 
-------
2 : protein_gtr.rev.mle=estimators.FitSingleModel_Ext(utility.Map(filter_info,"_value_","_value_['name']"),trees,protein_gtr.rev_model_branch_lengths,initial_values,{"proportional-branch-length-scaler":proportional_scalers})
-------
3 : protein_gtr.current_gtr_fit=protein_gtr.fitGTRtoFileList(utility.Map(utility.Filter(protein_gtr.analysis_results,"_value_","_value_/'"+protein_gtr.phase_key+"'"),"_value_","_value_['"+protein_gtr.phase_key+"']"),protein_gtr.current_gtr_fit,result_key,FALSE)
```

Files:
[errors.log.mpinode0.txt](https://github.com/veg/hyphy/files/880049/errors.log.mpinode0.txt)
[messages.log.mpinode0.txt](https://github.com/veg/hyphy/files/880050/messages.log.mpinode0.txt)

**Comment by spond** [comment]
Attempting to replicate on a clean build

**Comment by spond** [comment]
@sjspielman : did you turn on rate variation for this "buggy" run?

**Comment by sjspielman** [comment]
@spond, Yes, there was rate variation. Trying now without.

**Comment by spond** [comment]
@sjspielman: OK error replicated with rate variation. Suspect it has to do with MPI-enabled optimization in the presence of rate variation (hasn't been tested in a while)

**Comment by sjspielman** [comment]
I have also replicated the error in the absence of rate variation, so while possibly rate variation could still cause problems, it doesn't appear to be the culprit here.

**Comment by spond** [comment]
@sjspielman : OK trying to sandbox the issue (it's a weird internal bug). Stay tuned. This is useful -- I like squashing internal core (edge-case) bugs.

**Comment by spond** [comment]
Dear @sjspielman,

Should be resolved via this v2.3-dev commit : e19f0d39355421c9cf1644fad8f63d34fa6c0c6b

Best,
Sergei
